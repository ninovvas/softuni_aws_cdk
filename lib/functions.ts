import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Construct} from "constructs";
import {Architecture, Runtime} from "aws-cdk-lib/aws-lambda";
import {Duration} from "aws-cdk-lib";


interface BaseFunctionProps {
    tableName: string;
    handlerLocation: string;
}

export class BaseFunction extends NodejsFunction {
    constructor(scope: Construct, id: string, props: BaseFunctionProps) {
        super(scope, id, {
            ...props,
            runtime: Runtime.NODEJS_20_X,
            handler: 'handler',
            entry: `${__dirname}/../src/${props.handlerLocation}.ts`,
            timeout: Duration.seconds(10),
            architecture: Architecture.ARM_64,
            environment: {
                TABLE_NAME: props.tableName
            }

        });
    }
}