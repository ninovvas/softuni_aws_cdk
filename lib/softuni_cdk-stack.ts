import * as cdk from 'aws-cdk-lib';
import {CfnOutput, Duration} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {Bucket} from "aws-cdk-lib/aws-s3";
import {AttributeType, BillingMode, Table} from "aws-cdk-lib/aws-dynamodb";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Architecture, Runtime} from "aws-cdk-lib/aws-lambda"
import {BaseFunction} from "./functions";

//import * as sqs from 'aws-cdk-lib/aws-sqs';

export class SoftuniCdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here

   /* const queue: cdk.aws_sqs.Queue = new Queue(this, 'SoftuniCdkQueue', {
      visibilityTimeout: cdk.Duration.seconds(300),
      queueName: 'ExampleQueue',
      removalPolicy: RemovalPolicy.RETAIN
    });

    const lamdaFunction :cdk.aws_lambda.Function = new Function(this, 'LamdaWithGenericConstruct', {
      runtime: Runtime.NODEJS_20_X,
      handler: 'generic.handler',
      code: Code.fromAsset(`${__dirname}/../src`)
    })*/

    // example resource
    //const queue = new sqs.Queue(this, 'SoftuniCdkQueue', {
    //   visibilityTimeout: cdk.Duration.seconds(300)
    // });


    const exampleBucket :cdk.aws_s3.Bucket = new Bucket(this, 'ExampleBucket');

    const exampleTable :cdk.aws_dynamodb.Table = new Table(this, 'ExampleTable', {
      partitionKey: {
        name: 'PK',
        type: AttributeType.STRING
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
    });

    const baseFunction :cdk.aws_lambda_nodejs.NodejsFunction = new NodejsFunction(this, 'BaseFunction', {
      runtime: Runtime.NODEJS_20_X,
      handler: 'handler',
      entry: `${__dirname}/../src/generic.ts`,
      timeout: Duration.seconds(10),
      architecture: Architecture.ARM_64,
      environment: {
        TABLE_NAME: exampleTable.tableName
      }

    })

  const baseFunctionOne :cdk.aws_lambda_nodejs.NodejsFunction = new BaseFunction(this, 'one', {
      handlerLocation: 'one',
      tableName: exampleTable.tableName

    })

    const baseFunctionTwo :cdk.aws_lambda_nodejs.NodejsFunction = new BaseFunction(this, 'two', {
      handlerLocation: 'two',
      tableName: exampleTable.tableName

    })

    new CfnOutput(this, 'tableName', {
      value: exampleTable.tableName,
    });

    new CfnOutput(this, 'bucketName', {
      value: exampleBucket.bucketName
    })

  }
}
