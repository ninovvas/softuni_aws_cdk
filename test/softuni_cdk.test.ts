 import * as cdk from 'aws-cdk-lib';
 import { Template } from 'aws-cdk-lib/assertions';
 import * as SoftuniCdk from '../lib/softuni_cdk-stack';
 import 'jest-cdk-snapshot';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/softuni_cdk-stack.ts
test('SQS Queue Created', () => {
   const app = new cdk.App();
//     // WHEN
   const stack = new SoftuniCdk.SoftuniCdkStack(app, 'MyTestStack');
//     // THEN
   //const template = Template.fromStack(stack);
   expect(stack).toMatchCdkSnapshot();
});