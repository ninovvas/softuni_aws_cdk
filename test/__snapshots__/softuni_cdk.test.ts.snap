// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Queue Created 1`] = `
{
  "Outputs": {
    "bucketName": {
      "Value": {
        "Ref": "ExampleBucketDC717CF4",
      },
    },
    "tableName": {
      "Value": {
        "Ref": "ExampleTable114D508F",
      },
    },
  },
  "Resources": {
    "BaseFunction2720903A": {
      "DependsOn": [
        "BaseFunctionServiceRole9D84751B",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f35a6e179b1780f489f8e8542d227424ff0f31d1bb587b6755eea1ded64c610a.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "ExampleTable114D508F",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BaseFunctionServiceRole9D84751B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BaseFunctionServiceRole9D84751B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExampleBucketDC717CF4": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "ExampleTable114D508F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "one0A8EDDC8": {
      "DependsOn": [
        "oneServiceRole9B3089FC",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9041c640c3bb676a562eee4b768a345135f883529f3307f29c49514549551fcb.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "ExampleTable114D508F",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "oneServiceRole9B3089FC",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "oneServiceRole9B3089FC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "two9720DAB2": {
      "DependsOn": [
        "twoServiceRole056C1EB2",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a20b8017ab40fcf73e1ed32d0438ec75fe81228bdb7705b81288bc44f3175b1f.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "ExampleTable114D508F",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "twoServiceRole056C1EB2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "twoServiceRole056C1EB2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
